src/ni/vsbuild/Pipeline.groovy

```diff
@@ -173,16 +174,17 @@ class Pipeline implements Serializable {
       }
 
       jsonConfig = config.toString()
+      return manifest['scm']['GIT_COMMIT']
    }
 
-   private void validateShouldBuildPipeline() {
+   private def validateShouldBuildPipeline(commit) {
```

Comment:
This method now has three potential exit modes:

1. Return true, indicating we should build.
1. Return false, indicating we are reusing an existing build.
1. Fail the build (i.e. don't return anything), indicating something is wrong (like the output location is occupied).

Can you document this behavior? It's not immediately obvious when glancing at the code.

Reply:
I can document it if we keep it this wasy.

I debated putting the rebuild check in a different method, but was trying to avoid creating another `node()` acquisition because each time we do that, we run the risk of delaying the build further if there are no nodes available. I don't know how much of an issue that really is. It probably makes more sense to pull this out of `validateShouldBuildPipeline` and move it to a separate `checkIfRebuildNeeded`.

Thoughts?

Reply:
I agree that we should avoid calling `node()` more than needed.

Can we invert the relationship here, and have `node()` get called outside of `validateShouldBuildPipeline()` and `checkIfRebuildNeeded()`? Then they could share the reservation, while still (potentially) having their own stages. I'm not sure how feasible that is.

Reply:
Maybe. It'll take a little rework, but we could probably share the node and split the methods.